app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt

```diff
@@ -0,0 +1,224 @@
+        p.typeface = Typeface.MONOSPACE
+        p.textAlign = Paint.Align.RIGHT
+        p.textSize = textSize * 0.8f
+        p.flags = p.flags or Paint.ANTI_ALIAS_FLAG
+        p
+    }
+
+    var lang = ""
+        set(value) {
+            field = value
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    
+    var scrollContainer: ScrollView? = null
+        set(value) {
+            field = value
+            value?.let { container ->
+                container.viewTreeObserver.addOnScrollChangedListener {
+                    scrollPublisher.onNext(container.scrollY)
+                }
+
+                container.viewTreeObserver.addOnGlobalLayoutListener {
+                    scrollPublisher.onNext(container.scrollY)
+                }
+
+                scrollDisposable = scrollPublisher
+                    .debounce(SCROLL_DEBOUNCE_MS, TimeUnit.MILLISECONDS)
+                    .subscribeOn(Schedulers.computation())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe { spanPublisher.value?.let(this::updateHighlight) }
+            }
+        }
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+
+        compositeDisposable.add(
+                spanPublisher
+                        .subscribeOn(Schedulers.computation())
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .subscribe { updateHighlight(it) }
+        )
+
+        compositeDisposable.add(
+                highlightPublisher
+                        .debounce(INPUT_DEBOUNCE_MS, TimeUnit.MILLISECONDS)
+                        .map {
+                            parser.parse(lang, it.toString())
+                        }
+                        .subscribeOn(Schedulers.computation())
+                        .observeOn(Schedulers.computation())
+                        .subscribe(spanPublisher::onNext)
+        )
+
+        addTextChangedListener(this)
+        afterTextChanged(editableText)
+    }
+
+    override fun onDetachedFromWindow() {
+        removeTextChangedListener(this)
+        compositeDisposable.dispose()
+        scrollDisposable?.dispose()
+        super.onDetachedFromWindow()
+    }
+
+    var theme : CodeTheme = DefaultTheme
+        set(value) {
+            field = value
+            setBackgroundColor(value.background)
+            setTextColor(value.syntax.plain)
+            lineNumbersBackgroundPaint.color = value.lineNumberBackground
+            lineNumbersStrokePaint.color = value.lineNumberStroke
+            lineNumbersTextPaint.color = value.lineNumberText
+            selectedLinePaint.color = value.selectedLineBackground
+
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    private val bufferRect = Rect()
+
+    var lines: List<String> = emptyList()
+    var linesWithNumbers: List<Int> = emptyList()
+
+    override fun onDraw(canvas: Canvas) {
```

Comment:
`onDraw` allocates a lot of memory now. Please avoid it. http://prntscr.com/gwtu61